### DISCUSSION TITLE
PrismaClientValidationError with many to many relationships

### DISCUSSION THREAD

**Discussion by arthmrx** [prisma/prisma#5261] [Q&A] [answered]
Hello everyone,

I am trying to learn Prisma. In my demo app I wan to create a relationship between articles and categories.

```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model Article {
  id         Int        @id @default(autoincrement())
  title      String     @unique
  content    String?
  author     User?      @relation(fields: [authorId], references: [id])
  authorId   Int?
  categories Category[]
  createdAt  DateTime   @default(now()) @map(name: "created_at")
  updatedAt  DateTime   @updatedAt @map(name: "updated_at")

  @@map(name: "articles")
}

model User {
  id        Int       @id @default(autoincrement())
  name      String    @unique
  email     String    @unique
  password  String
  createdAt DateTime  @default(now()) @map(name: "created_at")
  updatedAt DateTime  @updatedAt @map(name: "updated_at")
  articles  Article[]
  image     String?

  @@map(name: "users")
}

model Category {
  id       Int       @id @default(autoincrement())
  name     String    @unique
  articles Article[]

  @@map(name: "categories")
}
```

```ts
const body = req.body;
      try {
        const article = await prisma.article.create({
          data: {
            title: body.title,
            content: body.content,
            authorId: body.userId,
            categories: {
              create: {
                name: body.category
              }
            }
          }
        });
        res.status(200).json(article);
      } catch (err) {
        console.log(err);
        res.status(400).send(err);
      }
```

For some reason I always get this error when sending my request:

```
PrismaClientValidationError:
Invalid `prisma.article.create()` invocation:

{
  data: {
    title: 'Second article',
    content: 'Blah Blah 2',
    authorId: 1,
    categories: {
    ~~~~~~~~~~
      create: 
        {
          name: 'Cars'
        }
    }
  }
}

Unknown arg `categories` in data.categories for type ArticleUncheckedCreateInput. Available args:

type ArticleUncheckedCreateInput {
  id?: Int
  title: String
  content?: String | Null
  authorId?: Int | Null
  createdAt?: DateTime
  updatedAt?: DateTime
}
```

From my understanding this is basically the same as this support article: https://www.prisma.io/docs/support/help-articles/working-with-many-to-many-relations

I already tried to delete my db on heroku and then:

```
npx prisma db push --preview-feature
npx prisma generate
```

...but the request still fails with the same error.

**Comment by ryands17** [comment, marked as answer]
Hey @arthmrx 👋 
You need to use `connect` for the author in this case as that's the way you would need to add the author:

```ts
  const article = await prisma.article.create({
    data: {
      title: 'title',
      content: 'content',
      author: {
        connect: {
          id: 1,
        },
      },
      categories: {
        create: {
          name: 'category',
        },
      },
    },
  })
```

Also you are referring to the doc for working with many-to-many relations which is not relevant in your case. You can read about working with relations [here](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-writes).

**Comment by arthmrx** [comment]
Hi @ryands17,

thanks for the reply. The author works just fine with your code and also with my code (docs say `In 2.11.0 and later, you can set the foreign key directly`). So if I remove the categories part it works. However I want to an article to have a category associated with it and this is where the error message comes into play.

I dont unterstand why it complains about `Unknown arg "categories" in data.categories for type ArticleUncheckedCreateInput. Available args` even though my article model has a categories field.

**Reply by ryands17** [reply]
You can set the foreign key directly, but cannot mix the foreign key and connect mechanisms together. You have to use either one or the other which is why you're facing this issue.

If you need to use `create` or `connect` for `categories`, you would need to do the same for `author` as well. Hope that helps :)

**Comment by arthmrx** [comment]
That was indeed the mistake @ryands17. After I changed the author query to use `connect` it worked like a charm. 
Thanks for your help :hugs:

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
